### DISCUSSION TITLE
How do I get doReturn/when rather than when/thenReturn?

### DISCUSSION THREAD

**Discussion by dirkniblick** [SquaretestLLC/Squaretest#28] [General]
I am using Squaretest to generate tests in my Spring Boot project via IntelliJ. I want my tests to contain mocks of the form...

`doReturn(1).when(object).returnOne();`

...rather than...

`when(object.returnOne()).thenReturn(1);`

I'm currently using the JUnit5MockitoSpringAssertJ.java.ft template, but cannot figure out how to configure it to do this. The relevant portion of the template seems to look like this...

```
#set($macro.hasInnerWildcardType = $di.method.returnType.parameters.containsAnyWith('wildcard', true))

...

#if($macro.hasInnerWildcardType)
    $doReturn($macro.returnTypeExpression).$when($macro.qualifier).#methodCall($di.method, false);
#else
    $when($macro.qualifier.#methodCall($di.method, false)).$thenReturn($macro.returnTypeExpression);
#end
#if($macro.shouldRenderComment) $Newline #end
```

...but the setting of `hasInnerWildcardType` seems to be dependent on the method being tested...? I just realized, for my purposes, I can simply edit the template to remove the `if` completely. However, I feel you should have some sort of setting variable in the `if` condition which overrides `hasInnerWildcardType`...

```
#set($useDoReturn = true)

...

#set($macro.hasInnerWildcardType = $di.method.returnType.parameters.containsAnyWith('wildcard', true))

...

#if($useDoReturn || $macro.hasInnerWildcardType)
    $doReturn($macro.returnTypeExpression).$when($macro.qualifier).#methodCall($di.method, false);
#else
    $when($macro.qualifier.#methodCall($di.method, false)).$thenReturn($macro.returnTypeExpression);
#end
#if($macro.shouldRenderComment) $Newline #end
```

**Comment by nate-dev** [comment]
Thank you for the feedback.

> the setting of hasInnerWildcardType seems to be dependent on the method being tested...?

It is dependent on the dependency method (the method for which the when(..) statement is created). If the dependency method's return type has a wildcard, we need to use doReturn(..) to avoid getting a compiler error.

Templates can be updated using the logic you described to render doReturn(..) statements. You will also need to update similar code in lines: 1508, 1543, 1578, 1617 and 1656. I will look into adding a flag for this setting. However I believe this usecase may be too uncommon to warrant one at this time.

**Reply by dirkniblick** [reply]
Thanks for explaining that for me. I understand it might make extra work to add a flag, but I’m guessing there are others like me who prefer the style of the doReturn/when format over the other. I did add a flag to my template and used it for the doThrow lines too.

Your plugin makes writing tests quite a bit easier and I’m glad my boss bought me license. Thanks again.
